File: controller/effect.cpp
//
// Created by Nico Wittig on 2018-12-14.
//

#include "effect.h"


effect::effect(light_element *le, char *name) {
    this->le = le;
    this->name = name;
}


effect::~effect() {
}


char* effect::getName() {
    return this->name;
}

// region effect helper

bool effect::check_color_matching(CHSV *color, uint16_t begin_led, uint16_t end_led) {

    // Catch exceptions
    if (begin_led > end_led) { // first led after second -> switch leds
        uint16_t led = begin_led;
        begin_led = end_led;
        end_led = led;
    }

    for (uint16_t led = begin_led; led <= end_led; led++) {
        if (this->le->leds[led] != *color) return false;
    }

    return true;
}


bool effect::check_color_matching(CHSV *color) {
    return check_color_matching(color, 0, le->get_num_leds()-1);
}


CHSV effect::get_pixel_gradient(CHSV begin_color, CHSV end_color, bool hue_cycle_forward, uint16_t area_length,
                                uint16_t index_led) {

    // calculate variable values
    float diff_sat = (float) (end_color.sat-begin_color.sat)/(area_length-1);
    float diff_val = (float) (end_color.val-begin_color.val)/(area_length-1);
    float diff_hue;

    if (hue_cycle_forward) {
        diff_hue = (float) ((uint8_t) (end_color.hue-begin_color.hue))/(area_length-1);
    } else {
        diff_hue = (float) (-(uint8_t) (begin_color.hue-end_color.hue))/(area_length-1);
    }

    return CHSV(
            begin_color.hue+index_led*diff_hue,
            begin_color.sat+index_led*diff_sat,
            begin_color.val+index_led*diff_val);
}

//region color setter

void effect::set_pixel(uint16_t pixel, CHSV color) {
    this->le->leds[pixel] = color;
}


void effect::set_pixel(uint16_t pixel, CRGB color) {
    this->le->leds[pixel] = rgb2hsv_approximate(color);
}


void effect::set_color(CHSV begin_color, CHSV end_color, bool hue_cycle_forward, uint16_t begin_led, uint16_t end_led) {

    // Catch exceptions
    if (begin_led == end_led) { // only one led -> set to begin color
        le->leds[begin_led] = begin_color;
        return;
    } else if (begin_led > end_led) { // first led after second -> switch leds
        uint16_t led = begin_led;
        begin_led = end_led;
        end_led = led;
    }

    // calculate variable values
    uint16_t area_length = end_led-begin_led+1;

    // set leds to new colors
    for (uint16_t i = 0; i < area_length; i++) {
        uint16_t current_led = begin_led + i;
        le->leds[current_led] = get_pixel_gradient(begin_color, end_color, hue_cycle_forward, area_length, i);
    }

}


void effect::set_color(CHSV begin_color, CHSV end_color, bool hue_cycle_forward, uint16_t begin_led, uint16_t end_led,
                       uint8_t cycles, bool forward_only) {

    // calculate variable values
    int area_length = end_led-begin_led+1;
    float cycle_length = (float) area_length/cycles;

    // call set_color for all cycles
    for (uint8_t c = 0; c < cycles; c++) {
        if (forward_only || c%2 == 0) {
            set_color(begin_color, end_color, hue_cycle_forward, begin_led+c*cycle_length, begin_led+(c+1)*cycle_length-1);
        } else {
            set_color(end_color, begin_color, !hue_cycle_forward, begin_led+c*cycle_length, begin_led+(c+1)*cycle_length-1);
        }
    }

}


void effect::set_color(CHSV begin_color, CHSV end_color, bool hue_cycle_forward, uint16_t begin_led, uint16_t end_led,
                       uint8_t cycles) {
    set_color(begin_color, end_color, hue_cycle_forward, begin_led, end_led, cycles, true);
}


void effect::set_color(CHSV begin_color, CHSV end_color, bool hue_cycle_forward) {
    set_color(begin_color, end_color, hue_cycle_forward, 0, le->get_num_leds()-1);
}


void effect::set_color(CHSV begin_color, CHSV end_color) {
    set_color(begin_color, end_color, true);
}


void effect::set_color(CHSV color) {
    set_color(color, color);
}

//endregion

void effect::increase_hue(int16_t value) {
    for (uint16_t led = 0; led < this->le->get_num_leds(); led++) {
        le->leds[led].hue += value;
    }
}


void effect::shift(int16_t direction) {
    if (direction > 0) {
        CHSV save_color = this->le->leds[this->le->get_num_leds() -1];

        for (uint16_t led = this->le->get_num_leds() -1; led >= 1; led--) {
            this->le->leds[led] = this->le->leds[led-1];
        }

        this->le->leds[0] = save_color;

        shift(--direction);
    } else if (direction < 0) {
        CHSV save_color = this->le->leds[0];

        for (uint16_t led = 0; led < this->le->get_num_leds(); led++) {
            this->le->leds[led] = this->le->leds[led+1];
        }

        this->le->leds[this->le->get_num_leds()-1] = save_color;

        shift(++direction);
    } else {
        return;
    }
}


void effect::shift_hue(int16_t direction) {

    if (direction > 0) {
        uint8_t save_hue = le->leds[le->get_num_leds() -1].hue;

        for (int led = le->get_num_leds() -1; led >= 1; led--) {
            le->leds[led].hue = le->leds[led-1].hue;
        }

        le->leds[0].hue = save_hue;
        shift_hue(--direction);
    } else if (direction < 0) {
        uint8_t save_hue = le->leds[0].hue;

        for (int led = 0; led < le->get_num_leds(); led++) {
            le->leds[led].hue = le->leds[led+1].hue;
        }

        le->leds[le->get_num_leds()-1].hue = save_hue;
        shift_hue(++direction);
    } else {
        return;
    }

}


bool effect::wipe(CHSV begin_color, CHSV end_color, bool hue_cycle_forward, uint16_t begin_led, uint16_t end_led,
                  int8_t direction) {

    if (begin_led > end_led) { // first led after second -> switch leds
        uint16_t led = begin_led;
        begin_led = end_led;
        end_led = led;
    }

    if (direction == 0) {
        return true;
    } else {

        // calculate variable values
        uint16_t area_length = abs(end_led-begin_led)+1;
        uint16_t max = 0;
        uint16_t min = 0u-1;

        for (uint16_t i = begin_led; i <= end_led; i++) {
            if (le->leds[i] != get_pixel_gradient(begin_color, end_color, hue_cycle_forward, area_length, i)) {
                if (i < min) min = i;
                if (i > max) max = i;
            }
        }

        if (direction > 0) {
            //le->leds[min] = new_color;
            for (int16_t j = 0; j < direction && min+j <= end_led; j++)
                le->leds[min+j] = get_pixel_gradient(begin_color, end_color, hue_cycle_forward, area_length, min+j-begin_led);
        } else {
            //le->leds[max] = new_color;
            for (int16_t j = 0; j > direction && ((int16_t) max)+j >= ((int16_t) begin_led); j--)
                le->leds[max+j] = get_pixel_gradient(begin_color, end_color, hue_cycle_forward, area_length, max+j-begin_led);
        }

    }

    return false;
}


bool effect::wipe(CHSV new_color, uint16_t begin_led, uint16_t end_led, int8_t direction) {
    return wipe(new_color, new_color, true, begin_led, end_led, direction);
}


bool effect::wipe(CHSV new_color, int8_t direction) {
    return wipe(new_color, 0, le->get_num_leds()-1, direction);
}


bool effect::fade_var(uint8_t *var, uint8_t end, int8_t step) {
    //Serial.print(step);
    if (step == 0 || *var == end) return true;

    *var += step;

    /*
    if ((step > 0 && *var >= end && *var < (uint8_t) (end+step))
        || step < 0 && (*var <= end || *var > (uint8_t) (end+step))) {
    */
    if ((step > 0 && *var > (uint8_t) (end-step) && *var <= end)
        || (step < 0 && *var < (uint8_t) (end-step) && *var >= end)) {

        *var = end;
        return true;
    }

    return false;
}


bool effect::fade(CHSV new_color, uint16_t pixel, uint8_t step) {

    // calculate variable values
    bool pixel_ready = true;
    int diff = (int) new_color.hue - (int) le->leds[pixel].hue;
    bool pos_step = (abs(diff) <= 128) == (diff >= 0);

    // fade the hue, sat and val
    if (!fade_var(&le->leds[pixel].hue, new_color.hue, (pos_step) ? step : -step))
        pixel_ready = false;

    if (!fade_var(&le->leds[pixel].sat, new_color.sat,
                  (le->leds[pixel].sat > new_color.sat) ? -step : step))
        pixel_ready = false;

    if (!fade_var(&le->leds[pixel].val, new_color.val,
                  (le->leds[pixel].val > new_color.val) ? -step : step))
        pixel_ready = false;

    return pixel_ready;
}


bool effect::fade(CHSV begin_color, CHSV end_color, bool hue_cycle_forward, uint16_t begin_led, uint16_t end_led,
                  int8_t step) {

    // calculate variable values
    uint16_t area_length = end_led-begin_led+1;
    bool all_pixels_ready = true;

    for (int led = begin_led; led <= end_led; led++) {
        if (!fade(get_pixel_gradient(begin_color, end_color, hue_cycle_forward, area_length, led-begin_led), led, step)) all_pixels_ready = false;
    }

    return all_pixels_ready;
}


bool effect::fade(CHSV new_color, uint16_t begin_led, uint16_t end_led, int8_t step) {
    return fade(new_color, new_color, true, begin_led, end_led, step);
}


bool effect::fade(CHSV new_color, int8_t step) {
    return fade(new_color, 0, le->get_num_leds()-1, step);
}

// endregion

File: server/effect.py
import sys

from color import Color


class Effect(object):
    def __init__(self, light_element, name="Effect"):
        super().__init__()

        self.__light_element = light_element
        self.__name = name

    def __get_name(self):
        return self.__name

    name = property(__get_name)

    def __str__(self):
        return self.__name + ": " + str(self.__light_element)

    # returns the color of a given pixel
    def get_pixel(self, index):
        return self.__light_element.get_pixel(index)

    # set a certain pixel to a specific color
    def set_pixel(self, index, color):
        self.__light_element.set_pixel(index, color)

    # checks the attached light element if the colors are similar to the given color
    def check_color_matching(self, color, start_pixel=0, end_pixel=None):

        # no end pixel -> set to end of light element
        if not end_pixel:
            end_pixel = self.__light_element.num_pixels-1

        # first pixel after second -> switch
        if start_pixel > end_pixel:
            start_pixel, end_pixel = end_pixel, start_pixel

        for pixel_index in range(start_pixel, end_pixel+1):
            if self.get_pixel(pixel_index) != color:
                return False

        return True

    # HELPER: decide the hue cycle direction depending on used colors (shortest path)
    def __get_hue_cycle_forward(self, start_color, end_color):
        diff_hue = end_color.hue - start_color.hue
        #hue_cycle_forward = (diff_hue > 0 and diff_hue < 128) or (diff_hue < 0 and diff_hue <= -128)
        hue_cycle_forward = (0 < diff_hue < 128) or diff_hue <= -128
        return hue_cycle_forward

    # get the color the pixel would have if the given gradient would be applied
    def __get_pixel_gradient(self, index_pixel, start_color, end_color=None, hue_cycle_forward=None, area_length=None):

        # no end color -> set to start color
        if not end_color:
            end_color = start_color

        # decide the hue cycle direction depending on used colors (shortest path)
        if hue_cycle_forward == None:
            hue_cycle_forward = self.__get_hue_cycle_forward(start_color, end_color)

        area_length = abs(area_length) if area_length else self.__light_element.num_pixels

        diff_sat = (end_color.sat-start_color.sat)/(area_length-1)
        diff_val = (end_color.val-start_color.val)/(area_length-1)

        if hue_cycle_forward:
            diff_hue = ((end_color.hue-start_color.hue)%256)/(area_length-1)
        else:
            diff_hue = -((start_color.hue-end_color.hue)%256)/(area_length-1)

        return Color((
            start_color.hue+index_pixel*diff_hue,
            start_color.sat+index_pixel*diff_sat,
            start_color.val+index_pixel*diff_val
        ))

    # HELPER: set a gradient from the first to the second color on the specified area on the light element
    def __set_color(self, start_color, end_color=None, hue_cycle_forward=None, start_pixel=0, end_pixel=None):

        # no end pixel -> set to end of light element
        if not end_pixel:
            end_pixel = self.__light_element.num_pixels-1

        # only one pixel -> set to start color
        if start_pixel == end_pixel:
            self.set_pixel(start_pixel, start_color)
            return
        # first pixel after second -> switch
        elif start_pixel > end_pixel:
            start_pixel, end_pixel = end_pixel, start_pixel

        area_length = end_pixel-start_pixel+1

        for i in range(area_length):
            current_pixel = start_pixel + i
            self.set_pixel(current_pixel, self.__get_pixel_gradient(i, start_color, end_color, hue_cycle_forward, area_length))

    # set one or multiple gradients from the first to the second color on the specified area on the light element
    def set_color(self, start_color, end_color=None, hue_cycle_forward=None, start_pixel=0, end_pixel=None, cycles=1, forward_only=True):

        # no end color -> set to start color
        if end_color == None:
            end_color = start_color

        # no end pixel -> set to end of light element
        if end_pixel == None:
            end_pixel = self.__light_element.num_pixels-1

        # expect integers
        start_pixel = int(start_pixel)
        end_pixel = int(end_pixel)

        # first pixel after second -> switch
        if start_pixel > end_pixel:
            start_pixel, end_pixel = end_pixel, start_pixel

        # decide the hue cycle direction depending on used colors (shortest path)
        if hue_cycle_forward == None:
            hue_cycle_forward = self.__get_hue_cycle_forward(start_color, end_color)

        area_length = end_pixel-start_pixel+1
        cycle_length = area_length/cycles

        for c in range(cycles):
            if forward_only or c%2 == 0:
                self.__set_color(start_color, end_color, hue_cycle_forward, int(start_pixel+c*cycle_length), int(start_pixel+(c+1)*cycle_length-1))
            else:
                self.__set_color(end_color, start_color, not hue_cycle_forward, int(start_pixel+c*cycle_length), int(start_pixel+(c+1)*cycle_length-1))

    # increase the hue of every pixel of the attached light element by value
    def increase_hue(self, value):
        for p in self.__light_element.pixels:
            p.hue += value

    # shift all pixels of the attached light element by the given direction-value
    def shift(self, direction):
        if direction > 0:

            save_color = self.get_pixel(self.__light_element.num_pixels-1)

            for i in range(self.__light_element.num_pixels-1, 0, -1):
                self.set_pixel(i, self.get_pixel(i-1))
            self.set_pixel(0, save_color)

            self.shift(direction-1)

        elif direction < 0:

            save_color = self.get_pixel(0)

            for i in range(0, self.__light_element.num_pixels-1):
                self.set_pixel(i, self.get_pixel(i+1))
            self.set_pixel(self.__light_element.num_pixels-1, save_color)

            self.shift(direction+1)

        else:
            return

    # shift only the hue-values of all pixels of the attached light element by the given direction-value
    def shift_hue(self, direction):
        if direction > 0:

            save_hue = self.get_pixel(self.__light_element.num_pixels - 1).hue

            for i in range(self.__light_element.num_pixels - 1, 0, -1):
                self.get_pixel(i).hue = self.get_pixel(i - 1).hue
            self.get_pixel(0).hue = save_hue

            self.shift_hue(direction - 1)

        elif direction < 0:

            save_hue = self.get_pixel(0).hue

            for i in range(0, self.__light_element.num_pixels - 1):
                self.get_pixel(i).hue = self.get_pixel(i + 1).hue
            self.get_pixel(self.__light_element.num_pixels - 1).hue = save_hue

            self.shift_hue(direction + 1)

        else:
            return

    # wipe a gradient through the area between one led and another
    def wipe(self, direction, start_color, end_color=None, hue_cycle_forward=None, start_pixel=0, end_pixel=None):

        if direction == 0:
            return True

        # no end pixel -> set to end of light element
        if end_pixel == None:
            end_pixel = self.__light_element.num_pixels-1

        # first pixel after second -> switch
        if start_pixel > end_pixel:
            start_pixel, end_pixel = end_pixel, start_pixel

        area_length = end_pixel - start_pixel + 1
        min_pixel = sys.maxsize
        max_pixel = -sys.maxsize - 1
        done = True

        # get the first (and last) pixel not yet wiped
        for i in range(start_pixel, end_pixel + 1):
            if self.get_pixel(i) != self.__get_pixel_gradient(i, start_color, end_color, hue_cycle_forward, area_length):
                if i < min_pixel:
                    min_pixel = i
                if i < min_pixel:
                    max_pixel = i
                done = False

        # all pixels already wiped
        if done:
            return True

        if direction > 0:
            i = 0
            while i < direction and min_pixel + i <= end_pixel:
                self.set_pixel(min_pixel + i, self.__get_pixel_gradient(min_pixel + i - start_pixel, start_color, end_color, hue_cycle_forward, area_length))
                i += 1
        else:
            i = 0
            while i > direction and max_pixel + i >= start_pixel:
                self.set_pixel(max_pixel + i, self.__get_pixel_gradient(max_pixel + i - start_pixel, start_color, end_color, hue_cycle_forward, area_length))
                i -= 1

        return False

    # HELPER: fade one single parameter to a certain value by the given steps (one step at a time)
    # returns a tuple: goal-value reached (True/False) and the value
    def __fade_var(self, val, end, step):
        diff = (end-val)%256
        if (step == 0 or val == end         # end value already reached -> set to end
        or (step > 0 and diff < step)       # less than one step is neaded -> set to end
        or (step < 0 and diff-256 > step)):
            return (True, end%256)

        val += step

        diff = (end-step)%256
        if ((step > 0 and val > diff and val <= end)   # no next steps needed -> set to end
         or (step < 0 and val < diff and val >= end)):
            return (True, end%256)

        return (False, val%256)

    # HELPER: fade one pixel of the attached light element to a certain color
    def __fade(self, step, new_color, pixel, short_fade=False):
        current_pixel = self.get_pixel(pixel)

        # short fade uses step or -step regarding to the shortest path for hue
        diff_hue = new_color.hue - current_pixel.hue
        pos_step = (abs(diff_hue) <= 128) == (diff_hue >= 0)

        fade_hue = self.__fade_var(current_pixel.hue, new_color.hue, step if not short_fade or pos_step else -step)
        current_pixel.hue = fade_hue[1]

        fade_sat = self.__fade_var(current_pixel.sat, new_color.sat, step if current_pixel.sat <= new_color.sat else -step)
        current_pixel.sat = fade_sat[1]

        fade_val = self.__fade_var(current_pixel.val, new_color.val, step if current_pixel.val <= new_color.val else -step)
        current_pixel.val = fade_val[1]

        return fade_hue[0] and fade_sat[0] and fade_val[0]

    # fade all pixels in an area of the attached light element to a certain color gradient
    def fade(self, step, start_color, end_color=None, hue_cycle_forward=None, start_pixel=0, end_pixel=None, short_fade=False):

        # no end pixel -> set to end of light element
        if end_pixel == None:
            end_pixel = self.__light_element.num_pixels-1

        area_length = end_pixel - start_pixel + 1
        all_pixels_ready = True

        for i in range(start_pixel, end_pixel + 1):
            if not self.__fade(step, self.__get_pixel_gradient(i - start_pixel, start_color, end_color, hue_cycle_forward, area_length), i, short_fade):
                all_pixels_ready = False

        return all_pixels_ready
